src/stream/src/executor/mview/materialize.rs

```diff
@@ -71,6 +74,145 @@ pub struct MaterializeExecutor<S: StateStore, SD: ValueRowSerde> {
     /// No data will be written to hummock table. This Materialize is just a dummy node.
     /// Used for APPEND ONLY table with iceberg engine. All data will be written to iceberg table directly.
     is_dummy_table: bool,
+
+    /// Optional refresh arguments and state for refreshable materialized views
+    refresh_args: Option<RefreshableMaterializeArgs<S, SD>>,
+}
+
+/// Arguments and state for refreshable materialized views
+pub struct RefreshableMaterializeArgs<S: StateStore, SD: ValueRowSerde> {
+    /// Table catalog for main table
+    pub table_catalog: Table,
+
+    /// Table catalog for staging table
+    pub staging_table_catalog: Table,
+
+    /// Flag indicating if this table is currently being refreshed
+    pub is_refreshing: bool,
+
+    /// During data refresh (between `RefreshStart` and `LoadFinish`),
+    /// data will be written to both the main table and the staging table.
+    ///
+    /// The staging table is PK-only.
+    ///
+    /// After `LoadFinish`, we will do a `DELETE FROM main_table WHERE pk NOT IN (SELECT pk FROM staging_table)`, and then purge the staging table.
+    pub staging_table: StateTableInner<S, SD>,
+
+    /// Table ID for this refreshable materialized view
+    pub table_id: TableId,
+}
+
+impl<S: StateStore, SD: ValueRowSerde> RefreshableMaterializeArgs<S, SD> {
+    /// Create new `RefreshableMaterializeArgs`
+    pub async fn new(
+        store: S,
+        table_catalog: &Table,
+        staging_table_catalog: &Table,
+        vnodes: Option<Arc<Bitmap>>,
+    ) -> Self {
+        let table_id = TableId::new(table_catalog.id);
+
+        // staging table is pk-only, and we don't need to check value consistency
+        let staging_table = StateTableInner::from_table_catalog_inconsistent_op(
+            staging_table_catalog,
+            store,
+            vnodes,
+        )
+        .await;
+
+        Self {
+            table_catalog: table_catalog.clone(),
+            staging_table_catalog: staging_table_catalog.clone(),
+            is_refreshing: false,
+            staging_table,
+            table_id,
+        }
+    }
+
+    /// `DELETE FROM original_table WHERE pk NOT IN (SELECT pk FROM tmp_table)`
+    pub async fn on_load_finish(
+        &mut self,
+        state_table: &mut StateTableInner<S, SD>,
+    ) -> StreamExecutorResult<()> {
```

Comment:
I think we should track the progress of on_load_finish and finish it in multiple barriers instead of one barrier. Otherwise, it would cause back-pressure.